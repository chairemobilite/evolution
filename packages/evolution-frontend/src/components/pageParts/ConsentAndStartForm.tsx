/*
 * Copyright 2023, Polytechnique Montreal and contributors
 *
 * This file is licensed under the MIT License.
 * License text available at https://opensource.org/licenses/MIT
 */
import React from 'react';

import { _isBlank } from 'chaire-lib-common/lib/utils/LodashExtensions';
import { WithTranslation, withTranslation } from 'react-i18next';
import config from 'chaire-lib-common/lib/config/shared/project.config';
import FormErrors from 'chaire-lib-frontend/lib/components/pageParts/FormErrors';

export type ConsentAndStartFormProps = {
    afterClicked: () => void;
};

type AgreementCheckboxProps = {
    text: string;
    onChange: (isChecked: boolean) => void;
};

const AgreementCheckbox = (props: AgreementCheckboxProps) => {
    const [checked, setChecked] = React.useState(false);
    const [updateKey, setUpdateKey] = React.useState(0);
    const onChange = (_e) => {
        const newChecked = !checked;
        setChecked(newChecked);
        props.onChange(newChecked);
    };

    return (
        <React.Fragment>
            <input
                type="checkbox"
                id="surveyConsent"
                key={`surveyConsentIdx${updateKey}`}
                className={'_input-checkbox'}
                value="agree"
                checked={checked}
                onChange={onChange}
            />
            <label
                htmlFor="surveyConsent"
                onClick={() => {
                    onChange;
                    setUpdateKey(updateKey + 1);
                }}
            >
                <span>{props.text}</span>
            </label>
        </React.Fragment>
    );
};

const ConsentAndStartForm = (props: ConsentAndStartFormProps & WithTranslation) => {
    const [agreed, setAgreed] = React.useState(false);
    const [showError, setShowError] = React.useState(false);
    const color = config.startButtonColor ? config.startButtonColor : 'green';
    const agreementText = props.t(['survey:homepage:AgreementText', 'main:AgreementText']);
    console.log('agreement text', agreementText);
    React.useEffect(() => {
        if (_isBlank(agreementText)) {
            setAgreed(true);
        }
    }, [agreementText]);
    const onButtonClicked = () => {
        if (!agreed) {
            setShowError(true);
        } else {
            props.afterClicked();
        }
    };
    const onAgreementChange = (hasAgreed) => {
        setShowError(false);
        setAgreed(hasAgreed);
    };
    return (
        <div className="survey-section__button-agreement">
            {!_isBlank(agreementText) && <AgreementCheckbox text={agreementText} onChange={onAgreementChange} />}
            {showError && (
                <FormErrors
                    errors={[props.t(['survey:homepage:ErrorNotAgreed', 'main:ErrorNotAgreed'])]}
                    errorType="Error"
                />
            )}
            <div className="center">
                <button
                    type="button"
                    className={`survey-section__button button ${color} large`}
                    onClick={onButtonClicked}
                >
                    {props.t(['survey:homepage:start', 'auth:Start'])}
                </button>
            </div>
        </div>
    );
};

export default withTranslation()(ConsentAndStartForm);
