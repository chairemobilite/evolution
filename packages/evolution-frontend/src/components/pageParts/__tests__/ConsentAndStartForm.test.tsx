/*
 * Copyright 2023, Polytechnique Montreal and contributors
 *
 * This file is licensed under the MIT License.
 * License text available at https://opensource.org/licenses/MIT
 */
import React from 'react';
import TestRenderer from 'react-test-renderer';
import { mount } from 'enzyme';

import ConsentAndStartForm from '../ConsentAndStartForm';

// Mock the react-i18next to control the translated string (see )
let translatedString = '';
jest.mock('react-i18next', () => ({
    // this mock makes sure any components using the translate HoC receive the t function as a prop
    withTranslation: () => Component => {
      Component.defaultProps = { ...Component.defaultProps, t: (str: string | string[]) => typeof str === 'string' ? str : str.findIndex(s => s.includes('AgreementText')) !== -1 ? translatedString : str[0] };
      return Component;
    },
}));


describe('Render ConsentAndStartForm', () => {

    test('Without consent checkbox', () => {
        translatedString = '';
        const wrapper = TestRenderer.create(
            <ConsentAndStartForm
                afterClicked={jest.fn()}
            />
        );
        expect(wrapper).toMatchSnapshot();
    });

    test('With consent checkbox', () => {
        translatedString = 'I agree';
        const wrapper = TestRenderer.create(
            <ConsentAndStartForm
                afterClicked={jest.fn()}
            />
        );
        expect(wrapper).toMatchSnapshot();
    });
    
});


describe('Button click', () => {

    test('Without consent', () => {
        const afterClick = jest.fn();
        translatedString = '';
        const consentAndStartForm = mount( <ConsentAndStartForm
            afterClicked={afterClick}
        />);
    
        // Click on button and make sure it accepts the change
        const formButton = consentAndStartForm.find(`.survey-section__button`).at(0);
        formButton.simulate('click');

        // Check that the callback has been called
        expect(afterClick).toHaveBeenCalled();
    });

    test('With consent, not checked', () => {
        const afterClick = jest.fn();
        translatedString = 'I agree';
        const consentAndStartForm = mount( <ConsentAndStartForm
            afterClicked={afterClick}
        />);
    
        // Click on button, it should not agree
        const formButton = consentAndStartForm.find(`.survey-section__button`).at(0);
        formButton.simulate('click');
        

        // Make sure the callback was not called and the error message appears
        expect(afterClick).not.toHaveBeenCalled();
        consentAndStartForm.update();
        const errors = consentAndStartForm.find(`.apptr__form-errors-container`).at(0);
        expect(errors.length).toEqual(1);
    });

    test('With consent, checked', () => {
        const afterClick = jest.fn();
        translatedString = 'I agree';
        const consentAndStartForm = mount( <ConsentAndStartForm
            afterClicked={afterClick}
        />);
    
        // Check the checkbox
        const consentCheckbox = consentAndStartForm.find(`#surveyConsent`).at(0);
        consentCheckbox.simulate('change');
        consentAndStartForm.update();

        // Click on button and make sure it accepts the change
        const formButton = consentAndStartForm.find(`.survey-section__button`).at(0);

        // Check the period and validate the choices
        formButton.simulate('click');
        expect(afterClick).toHaveBeenCalled();
    });

});
