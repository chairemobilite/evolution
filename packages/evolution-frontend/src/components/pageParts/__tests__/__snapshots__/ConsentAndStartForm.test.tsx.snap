// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render ConsentAndStartForm With consent checkbox 1`] = `
<div
  className="survey-section__button-agreement"
>
  <input
    checked={false}
    className="_input-checkbox"
    id="surveyConsent"
    onChange={[Function]}
    type="checkbox"
    value="agree"
  />
  <label
    htmlFor="surveyConsent"
    onClick={[Function]}
  >
    <span>
      I agree
    </span>
  </label>
  <div
    className="center"
  >
    <button
      className="survey-section__button button green large"
      onClick={[Function]}
      type="button"
    >
      survey:homepage:start
    </button>
  </div>
</div>
`;

exports[`Render ConsentAndStartForm Without consent checkbox 1`] = `
<div
  className="survey-section__button-agreement"
>
  <div
    className="center"
  >
    <button
      className="survey-section__button button green large"
      onClick={[Function]}
      type="button"
    >
      survey:homepage:start
    </button>
  </div>
</div>
`;
